src/main/java/org/testng/Assert.java

```diff
@@ -152,7 +152,7 @@ private static boolean areNotEqualImpl(Object actual, Object expected) {
     if ((expected == null) && (actual == null)) {
       return false;
     }
-    if (expected == null ^ actual == null) {
+    if(((expected == null) && !actual.equals(expected)) ^ ((actual == null) && !expected.equals(actual))){
```

Comment:
To match the code style of the other code of this class:
```suggestion
    if (((expected == null) && !actual.equals(expected)) ^ ((actual == null) && !expected.equals(actual))) {
```